src/render/gif.rs

```diff
@@ -0,0 +1,277 @@
+use std::{
+    cell::RefCell,
+    fs::File,
+    rc::Rc,
+    sync::atomic::{AtomicUsize, Ordering},
+    time::{Duration, Instant},
+};
+
+use apex_hardware::FrameBuffer;
+
+use embedded_graphics::{
+    image::{Image, ImageRaw},
+    pixelcolor::BinaryColor,
+    prelude::Point,
+    Drawable,
+};
+use gif::Frame;
+
+static GIF_MISSING: &[u8] = include_bytes!("./../../assets/gif_missing.gif");
+
+static DISPLAY_HEIGHT: i32 = 40;
+static DISPLAY_WIDTH: i32 = 128;
+
+pub struct Gif {
+    stop: Point,
+    origin: Point,
+    decoded_frames: Vec<Vec<u8>>,
+    current_frame: AtomicUsize,
+    delays: Vec<u16>,
+    time_frame_last_update: Rc<RefCell<Instant>>,
+}
+
+impl Gif {
+    pub fn calculate_median_color_value(frame: &Frame, gif_height: i32, gif_width: i32) -> u8 {
+        //NOTE we're using the median to determine wether the pixel should be black or
+        // white
+
+        let mut colors = (0..=255).into_iter().map(|_| 0).collect::<Vec<u32>>();
+
+        let width = frame.width;
+        let height = frame.height;
+
+        let num_pixels = gif_width as u32 * gif_height as u32;
+
+        let pixels = &frame.buffer;
+
+        for y in 0..gif_height {
+            //if y is outside of the gif width
+            if y >= height as i32 {
+                continue;
+            }
+
+            //if y is outside of the screen
+            if y >= DISPLAY_HEIGHT {
+                continue;
+            }
+            for x in 0..gif_width {
+                //if x is outside of the gif width
+                if x >= width as i32 {
+                    continue;
+                }
+
+                //if x is outside of the screen
+                if x >= DISPLAY_WIDTH {
+                    continue;
+                }
+
+                //calculating the index
+				// The u64 is just in case someone put a gif that's huge (in terms of
+				// resolution, on the x axis at least, since it's the only one that can 
+				// affect the index), if not, it will break (tested with a 1000*1000px gif)
+                let start = ((y as u64) * width as u64 + (x as u64)) * 4;
+
+                //getting the value of the pixels
+                let pixel_r = pixels.get(start as usize).unwrap_or(&0);
+                let pixel_g = pixels.get((start + 1) as usize).unwrap_or(&0);
+                let pixel_b = pixels.get((start + 2) as usize).unwrap_or(&0);
+                let pixel_a = pixels.get((start + 3) as usize).unwrap_or(&0);
```

Comment:
I just checked it out and you're right, this works but it still looks odd to me...

I came up with this but I'm not entirely sure whether its functionally equivalent:

```rust
let start = ((y as u64) * width as u64 + (x as u64)) * 4;

if let Some([r, g, b, a, ..]) = pixels.get(start as usize..) {
    let avg_pixel_value = ((*r as u32 + *g as u32 + *b as u32) / 3) as usize;
    colors[avg_pixel_value] += u32::from(*a) / 255;
}
```

Reply:
Better than `as u32` would be `u32::from` here